samwise.js

```diff
@@ -10,99 +10,104 @@
     "use strict";
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?